### PR TITLE
Fix harbinger timeout bug and _multipleFileReader docstring (resolves #953, resolves #949)

### LINKED ISSUE #949: AbstractCacheTest._multipleFileReader docstring is incorrect

### LINKED ISSUE #953: Caching hangs on harbinger file in sortTest
http://jenkins.cgcloud.info/job/toil-pull-requests/1413/

strace of the worker process

```
Trace of process 10637 - /mnt/ephemeral/workspace/toil-pull-requests/venv/bin/python /mnt/ephemeral/workspace/toil-pull-requests/venv/bin/_toil_worker /mnt/ephemeral/tmp/toil-test-toil.test.utils.utilsTe
Process 10637 attached
select(0, NULL, NULL, NULL, {2, 824883}) = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
select(0, NULL, NULL, NULL, {20, 0})    = 0 (Timeout)
flock(11, LOCK_EX)                      = 0
stat("/mnt/ephemeral/tmp/toil-43132b5e-c11d-4a83-aae6-8d302f973781/cache-43132b5e-c11d-4a83-aae6-8d302f973781/.TS9kL3RtcF80N1dGUC50bXA=.harbinger", {st_mode=S_IFREG|0664, st_size=0, ...}) = 0
flock(11, LOCK_UN)                      = 0
```

That's all I have.

### PR DESCRIPTION
**Pull request by arkal**
Resolves #953 
Resolves #949 

Harbinger timeout bug was fixed by adding logic for jobs to ensure that the process downloading the
requested file from the job store  is still alive, and to handle the case where it isn't.

Docstring in AbstractCacheTest._multipleFileReader was updated to reflect the changes made to it since
the function was written.

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
`Resolves` should be lowercase in first commit message.

File: src/toil/job.py [hunk 1]
```diff
@@ -1154,17 +1154,33 @@ def readGlobalFile(self, fileStoreID, userPath=None, cache=True, mutable=None):
...
-                        flock(lockFileHandle, LOCK_UN)
-                        time.sleep(20)  # What should this value be?
-                        flock(lockFileHandle, LOCK_EX)
+                        # Ensure that the process downloading the file is still alive.  The PID will
+                        # be in the harbinger file.
+                        try:
+                            pid = int(open(harbingerFileName).read())
+                        except ValueError:
+                            # There's a chance the downloading process died before it could write
+                            # into the harbinger file
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Then you have to create the file atomically. How do you know that it didn't write half of the pid which would still parse as an int?

File: src/toil/job.py [hunk 2]
```diff
@@ -1176,8 +1192,11 @@ def readGlobalFile(self, fileStoreID, userPath=None, cache=True, mutable=None):
                     logger.debug('CACHE: Cache miss on file with ID \'%s\'.' % fileStoreID)
                     if fileIsLocal and cache:
                         # If caching of the downloaded file is desired, First create the .partial
-                        # file so other jobs know not to redundantly download the same file.
-                        open(harbingerFileName, 'w').close()  # This emulates the system 'touch'
+                        # file so other jobs know not to redundantly download the same file.  Write
+                        # the PID of this process into the file so other jobs know who is carrying
+                        # out the download.
+                        with open(harbingerFileName, 'w') as harbingerFile:
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Yes, you need to create this atomically. We can't just reason "the lock is held, so I don't need to worry about atomicity" like we did earlier today.

File: src/toil/job.py [hunk 3]
```diff
@@ -1894,6 +1913,26 @@ def updateJobReqs(self, fileSize, actions):
                 # and it is added if the file was removed form the cache.
                 self.jobReqs -= (fileSize * multiplier)
 
+        @staticmethod
+        def _pidExists(pid):
+            """
+            This will return True if the process associated with pid is still running on the machine
+            Based on stackoverflow question 568271
```
Comment (hannes-ucsc) [inline src/toil/job.py]: period at the end of both sentences

File: src/toil/job.py [hunk 4]
```diff
@@ -1894,6 +1913,26 @@ def updateJobReqs(self, fileSize, actions):
...
+            :param int pid: Process ID
```
Comment (hannes-ucsc) [inline src/toil/job.py]: `ID of process to check for`

File: src/toil/job.py [hunk 5]
```diff
@@ -1894,6 +1913,26 @@ def updateJobReqs(self, fileSize, actions):
...
+            :return: True/False
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Redundant. But do

`:rtype: bool`

File: src/toil/job.py [hunk 6]
```diff
@@ -1894,6 +1913,26 @@ def updateJobReqs(self, fileSize, actions):
...
+                os.kill(pid, 0)
+            except OSError as err:
+                if err.errno == errno.ESRCH:
+                    # ESRCH == No such process
+                    return False
+                else:
+                    raise
+            else:
+                return True
+
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Great!

File: src/toil/job.py (hunk 1 above)
Comment (arkal) [inline src/toil/job.py]: good point

**Comment by arkal** [comment]
@hannes-ucsc i implemented the suggested changes

**Comment by hannes-ucsc** [comment]
Build failure tracked as #955.

File: src/toil/job.py (hunk 1 above)
Comment (hannes-ucsc) [inline src/toil/job.py]: Why is this still here? With the atomic creation you added, it is highly unlikely that you'll get a ValueError, correct? If so, I would rather let the ValueError through.

**Comment by hannes-ucsc** [comment]
@arkal, see line notes.

**Comment by arkal** [comment]
@hannes-ucsc i implemented the suggested changes

**Pull request merged**